### ISSUE TITLE
`@fastmath` interacts badly with complex

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#12167]
TL;DR
`@fastmath` somehow causes unnecessary allocations and slow down the calculations by 7000 times.

I was trying `@fastmath` in order to see if I can get, well...., faster math. To my surprise, I see a infinite times more allocations (since the code wasn't allocating) and a `7000x` times slow down because of this. A simplified version of my test which includes the benchmark both with and without `@fastmath` can be find [here](https://github.com/yuyichao/explore/blob/bdb00bd39141f4ea97d7ad7165ef6e9e1103fdc0/julia/fastmath/fastmath.jl). The function is basically:

```julia
function propagate(tmp, nstep)
    ψ_scale = 1.0
    p_x2 = 1.0
    for i in 1:nstep
        @fastmath tmp[1] = tmp[1] * p_x2 * ψ_scale
    end
end

propagate(zeros(Complex{Float64}, 1), 1)

@time propagate(zeros(Complex{Float64}, 1), 2560000)
```

On my laptop, the output of the script is

```julia
yuyichao% julia -f fastmath.jl
   1.429 seconds      (40960 k allocations: 1250 MB, 5.00% gc time)
 224.200 microseconds (9 allocations: 320 bytes)
```

where the first line is the `@fastmath` version and the second line is the one without it.

The relaxant part of the LLVM IR generated is

```llvm
  %17 = load %Complex.9** %13, align 8
  %18 = load %Complex.9* %17, align 8
  %19 = call %jl_value_t* @jl_gc_allocobj(i64 16)
  %20 = getelementptr inbounds %jl_value_t* %19, i64 -1, i32 0
  store %jl_value_t* inttoptr (i64 140728651152112 to %jl_value_t*), %jl_value_t** %20, align 8
  %21 = bitcast %jl_value_t* %19 to %Complex.9*
  store %Complex.9 %18, %Complex.9* %21, align 16
  store %jl_value_t* %19, %jl_value_t** %3, align 8
  store %jl_value_t* inttoptr (i64 140728687596256 to %jl_value_t*), %jl_value_t** %8, align 8
  store %jl_value_t* inttoptr (i64 140728687596272 to %jl_value_t*), %jl_value_t** %9, align 8
  %22 = call %jl_value_t* @julia_mul_fast_20807(%jl_value_t* inttoptr (i64 140728687416016 to %jl_value_t*), %jl_value_t** %3, i32 3)
```

So somehow the type inference decided not to inline the the multiplication and codegen decided to box the parameters and call it with the jlcall signature.

Probably not related to https://github.com/JuliaLang/julia/issues/11122 because it works fine with `Float64`, if I remove `p_x2` or if the results are not stored to an array. `@code_warntype` also doesn't show any instability (except for the `::Any` on explicit boxing).

Might be related to other inlining threshold issues?

**Comment by eschnett** [comment]
I think that fast-math complex multiplication does not explicitly handle the `Complex64 * Float64` case, whereas the regular multiplication does have a special case for this. This should lead to a type promotion (Float64 -> Complex64), followed by a complex multiplication -- but I see how a direct multiply-by-real would be faster.

See `base/fastmath.jl`; look for `mul_fast` there.

**Comment by eschnett** [comment]
Working on a fix...

**Comment by yuyichao** [comment]
Thx

**Comment by JeffBezanson** [comment]
I guess this must be related to the promoting `_fast` functions using varargs?

**Comment by JeffBezanson** [comment]
I notice the fastmath code also adds vararg promoting definitions for `abs` etc. Those shouldn't be necessary.

**Comment by eschnett** [comment]
@JeffBezanson I assume you mean the following: for `abs` etc., change these lines
```
        # fall-back implementation for non-numeric types
        $op_fast(xs...) = $op(xs...)
        # type promotion
        $op_fast(x::Number, y::Number, zs::Number...) =
            $op_fast(promote(x,y,zs...)...)
        # fall-back implementation that applies after promotion
        $op_fast{T<:Number}(x::T,ys::T...) = $op(x,ys...)
```
to these
```
        # fall-back implementation for non-numeric types
        $op_fast(xs...) = $op(xs...)
```

Do I understand this correctly?

**Comment by jakebolewski** [comment]
Closed by  #12182

**Comment by raminammour** [comment]
Hello all!

I noticed a problem with julia performace related to this issue. I was able to isolate it, maybe someone may help me understand why this happens, or maybe it can be fixed for everyone :)

It appears that @fastmath is fixed for float*complex, but not for float*float*complex, it still does the promotion to complex and so on, and is very slow. If one forces (float*float)*complex, it is fine (of course). The intriguing thing, is that without @fastmath, the problem does not arise!

Here is an example:
![image](https://cloud.githubusercontent.com/assets/20540368/17974344/5fa15c90-6aab-11e6-8a0c-0a55bedbb926.png)

Any help is much appreciated :)

**Comment by eschnett** [comment]
(Note: Instead of a screen shot, next time please post the actual code. Github will automatically colour it nicely if you mark it as Julia code. Also, Julia doesn't require semicolons at the end of a line.)

I think the problem is how a three-argument call to `fast_mul` is resolved. Such as call will first promote all arguments to the common type (here the complex type) and then perform the multiplications. If that is indeed the problem, then one way out would be to first hierarchically decompose multi-argument calls to `fast_mul` to two-argument calls, and only then apply type promotion.

We could even be fancy and use generated functions for this, so that we can determine the optimal multiplication order for any given set of types.

Same goes for `fast_add`.
